File: app/src/main/java/com/startai/winvkey/data_class/ObservableDataListObserver.java
package com.startai.winvkey.data_class;

import kotlin.reflect.KProperty;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;
import java.util.Observable;

/**
 * @Author: AiMin
 * @Description:
 * @Date: Created in 16:35 2018/11/16
 * @Modified By:
 */
public abstract class ObservableDataListObserver<E extends ObservableData> extends ObservableListObserver<E>{
    @Override
    public void update(Observable o, Object arg) {
        if(arg instanceof ObservableData.ChangeInfo) {
            try {
                ObservableData.ChangeInfo<E> castObj = (ObservableData.ChangeInfo) arg;
                this.itemChanged(castObj.getObj$app_debug(), castObj.getProperty$app_debug(), castObj.getOldVaule$app_debug(), castObj.getNewVaule$app_debug());
            } catch (ClassCastException e) {
                System.out.println("not a suitable arg for DataObservable: " + arg.getClass().getName());
            }
        }else{
            super.update(o, arg);
        }
    }

    public void itemChanged(E obj, @NotNull KProperty<?> property, @NotNull Object oldValue, @NotNull Object newValue){}
}

File: app/src/main/java/com/startai/winvkey/data_class/ObservableDataListObserver.kt
package com.startai.winvkey.data_class

import kotlin.reflect.KProperty

import java.lang.reflect.Field
import java.util.Observable

/**
 * @Author: AiMin
 * @Description:
 * @Date: Created in 16:35 2018/11/16
 * @Modified By:
 */
abstract class ObservableDataListObserver<E : ObservableData> : ObservableListObserver<E>() {
    override fun update(o: Observable, arg: Any) {
        if (arg is ObservableData.ChangeInfo<*>) {
            try {
                this.itemChanged(arg.obj as E, arg.property, arg.oldVaule, arg.newVaule)
            } catch (e: ClassCastException) {
                println("not a suitable arg for DataObservable: " + arg.javaClass.name)
            }

        } else {
            super.update(o, arg)
        }
    }

    open fun itemChanged(obj: E, property: KProperty<*>, oldValue: Any, newValue: Any) {}
}
